namedisl/__init__.py

```diff
@@ -32,72 +35,612 @@
+        return make_pw_multi_aff(self._obj.as_pw_multi_aff())
+
+    def eliminate(self, names: str | Sequence[str]) -> Self:
+        if isinstance(names, str):
+            names = [names]
+
+        new_obj = self._obj
+        for name in names:
+            dt, pos = self._name_to_dim[name]
+            new_obj = new_obj.eliminate(dt, pos, 1)
+
+        return type(self)(new_obj, self._name_to_dim)
+
+    def project_out(self, names_to_project_out: str | Sequence[str]) -> Self:
+        if isinstance(names_to_project_out, str):
+            names_to_project_out = [names_to_project_out]
+
+        new_name_to_dim = dict(self._name_to_dim)
+        new_isl_obj = self._obj
+        for proj_name in names_to_project_out:
+            dt, pos = new_name_to_dim[proj_name]
+            new_name_to_dim.pop(proj_name)
+
+            temp_name_to_dim = new_name_to_dim.copy()
+            for name, (cur_dt, cur_pos) in sorted(
+                    new_name_to_dim.items(), key=lambda k: k[1][1]):
+                if cur_pos > pos:
+                    temp_name_to_dim[name] = (cur_dt, cur_pos - 1)
+
+            new_name_to_dim = temp_name_to_dim.copy()
+
+            new_isl_obj = new_isl_obj.project_out(dt, pos, 1)
+
+        return type(self)(new_isl_obj, new_name_to_dim)
+
+    def project_out_except(self, names_to_keep: str | Sequence[str]) -> Self:
+        if isinstance(names_to_keep, str):
+            names_to_keep = [names_to_keep]
+
+        names_to_project_out = [
+            name for name in self._name_to_dim.keys()
+            if name not in names_to_keep
+        ]
+
+        return self.project_out(names_to_project_out)
+
+    def __and__(self, other) -> Self:
+        return _align_and_apply_op(self, other, operator.and_)
+
+    def __eq__(self, other) -> bool:
+        aligned_self, aligned_other = _align_two(self, other)
+        return aligned_self._obj.plain_is_equal(aligned_other._obj)  # type: ignore
+
+    def __or__(self, other) -> Self:
+        return _align_and_apply_op(self, other, operator.or_)
+
+    def __sub__(self, other) -> Self:
+        return _align_and_apply_op(self, other, operator.sub)
 
     def __str__(self) -> str:
-        return str(_restore_names(self._obj, self._name_to_dim))
+        return str(_restore_names(self))
+
+
+@dataclass(frozen=True, eq=False)
+class Set(_SetLike):
+    _obj: isl.Set
+
+
+@overload
+def make_set(src: str, ctx: isl.Context | None = None) -> Set:
+    ...
+
+
+@overload
+def make_set(src: isl.Set | isl.BasicSet) -> Set:
+    ...
+
+
+def make_set(
+        src: str | isl.Set | isl.BasicSet,
+        ctx: isl.Context | None = None
+    ) -> Set:
+    if isinstance(src, str):
+        obj = isl.Set(src, ctx)
+    elif isinstance(src, isl.BasicSet):
+        obj = isl.Set.from_basic_set(src)
+    else:
+        obj = src
+
+    obj, name_to_dim = _strip_names(obj)
+    return Set(obj, name_to_dim)
+
+
+@dataclass(frozen=True, eq=False)
+class Map(_SetLike):
+    _obj: isl.Map
```

Comment:
I don't think we should use `isl.Map` in a long-lived capacity; IMO, that distinction in isl does more harm than good in ISL (cf. all the places in loopy where there's back-and-forth conversion between maps and sets because the API on either end is incomplete.). I could see a `Map` data type that makes use of a namedisl `Set`.